### DISCUSSION TITLE
Are Tarred + bucketing data sets supposed to be faster?

### DISCUSSION THREAD

**Discussion by riqiang-dp** [NVIDIA-NeMo/NeMo#4217] [Q&A] [answered]
I'm trying to train ASR models on a pretty unbalanced data set (lengthwise, a lot short utterances but also long utterances (up to 1 min)). I have been able to train with smaller batch size and also with some pre-segmentation to shorten the max duration. I'm now just trying bucketing, but it seems like it only works with tarred data sets.

After tarring the original set and using some bucketing, I'm seeing that the estimated time for one epoch goes up to ~7h instead of 2h for training with the raw form audio files. If I don't do bucketing and still use the tarred version, I get ~10h estimated time for one epoch.

I'm setting something like 4 buckets, 128 shards. Using A100 GPUs on a Google Cloud VM, using NeMo 1.7.0

I'm just wondering, are these numbers normal? If so when does it make more sense to use tarred + bucketing strategy versus just train using the wav files?

**Comment by titu1994** [comment, marked as answer]
No this is not expected. At most a few seconds more is possible, not several times longer training. 

In the logs, can you check and plot the time per step (train and validation) between loose files and tarred / bucketed datasets? If it's not your model itself, then data loading in gcp is the problem. Which is odd because tarred datasets are what we use on such network filesystems.

**Reply by riqiang-dp** [reply]
I'm seeing step timing for the tarred dataset training run is around 0.65 vs non-tarred at 0.5. But it doesn't seem like that drastic of a difference...

**Reply by riqiang-dp** [reply]
It looks like I just didn't wait long enough? The actual epoch finished in 3 hours, but the progress bar only shows 50% when it's done, and it never got to evaluate the validation set...anything I might be doing wrong?

**Reply by riqiang-dp** [reply]
Just noticed these issues, seems to be related to this problem: https://github.com/NVIDIA/NeMo/issues/3854, https://github.com/NVIDIA/NeMo/issues/2810. Has there been a fix for ASR? Due to the wrong number of steps, no validation check, and no model checkpoints saved either.

**Reply by titu1994** [reply]
ASR model does not have those issues. Please check if your config is properly set when using tarred + bucketing. You should not set batch size > 1 and instead use bucketing batch size as a list. FYI @VahidooX  please add exact documentation or assertions to prevent such thing from allowing to execute during bucketing mode

**Reply by riqiang-dp** [reply]
It happens for tarred datasets with or without bucketing -> the epoch ends exactly at the 50% number of steps. 

There are already assertions to prevent incorrect batch settings so I should have already set it correctly: 
 ```
       model.train_ds.manifest_filepath=[[data/train/bucket1/tarred_audio_manifest.json],[data/train/bucket2/tarred_audio_manifest.json],[data/train/bucket3/tarred_audio_manifest.json],[data/train/bucket4/tarred_audio_manifest.json]] \
        +model.train_ds.tarred_audio_filepaths=[[data/train/bucket1/audio__OP_0..127_CL_.tar],[data/train/bucket2/audio__OP_0..127_CL_.tar],[data/train/bucket3/audio__OP_0..127_CL_.tar],[data/train/bucket4/audio__OP_0..127_CL_.tar]] \
        +model.train_ds.bucketing_strategy=synced_randomized \
        +model.train_ds.bucketing_batch_size=8 \
        model.train_ds.batch_size=1 \
        +model.train_ds.tarred_shard_strategy="scatter" \
        model.train_ds.shuffle_n=2048 \
        model.train_ds.is_tarred=True \
        model.train_ds.max_duration=50 \
        model.train_ds.min_duration=2.5 \
        model.train_ds.num_workers=32 \
        model.train_ds.pin_memory=True \
```

Now I'm thinking, could it be because I'm feeding in 8k wav and doing online upsampling that's slowing down the dataloading?  I see that training slows every 8 steps.

**Reply by titu1994** [reply]
That will slow it down but won't stop training after 50%

**Reply by riqiang-dp** [reply]
Right. Yeah I was talking about the speed regarding dataloading. Although I also trained with raw 8k wav files and the speed was much faster. 

For finishing an epoch early / wrong estimated step number, I don't know what's going on. 
What I used for creating the tarred dataset is this:
```
python <scripts path>/convert_to_tarred_audio_dataset.py \
        --manifest_path=data/train.json \
        --target_dir=data/train \
        --num_shards=128 \
        --max_duration=50 \
        --min_duration=2.5 \
        --shuffle \
        --shuffle_seed=1 \
        --sort_in_shards \
        --workers=32 \
        --buckets_num=4
```

Is it better if I open up an issue for this?

**Reply by VahidooX** [reply]
First, can you try reducing num_workers to a lower number like 8? 32 workers may cripple down your system if you have many GPUs on the same system.
Would you please share with us the number of samples in each buckets, total number of both the evaluation samples and training data, the batch size per GPU for train and eval, total number of GPUs, and the number of steps you see in the progress bar with both bucketing and non-bucketing datasets?

What was the largest batch size you could use without bucketing?
How could you use max_duration of 50s? There is no way to have a batch size of 8 with such large audios?
We suggest to use at most 20s for your training audio and split your long audios in shorter ones with the segmentation tools if you have so many.

**Reply by riqiang-dp** [reply]
Just tried reducing the num_workers, it doesn't seem to be affecting the speed at all. I saw somewhere that it's a good idea to set it 4*num_gpus, and I have 8 GPUs.

BTW the wrong number of steps per epoch turns out to be me using mismatched manifests / data (probably). But the slow training is real, for the same amount of data, it normally finishes one epoch with wav files in 30 mins, but it goes to over an hour with tarred + bucketing (segmented, max duration 20s), and if I use the non segmented files it's 4 hours (max duration 50s). 

> How could you use max_duration of 50s? There is no way to have a batch size of 8 with such large audios?

batch size of 8 seems to work with max_duration = 45s with wav files, but it seems to work for 50s max too if it's tarred and bucketed...probably because I use A100s. I have tried shortened audio, previously I found it giving worse results, but recently I found that it's probably because the splitting was not very good...for speed, training with shortened audio didn't affect training speed too much with raw wav file training, as long as I adapt the batch size, but shortened audio in tarred format trains much faster than unshortened..

The stats you asked for (might not be useful now since the steps number issue is resolved...but maybe still useful for why speed is bad?):
(and this is different data from the command line args I used above
```
bucket 1 with min_duration=2.5 and max_duration=6.875
172416 files
bucket 2 with min_duration=6.875 and max_duration=11.25
107136 files
bucket 3 with min_duration=11.25 and max_duration=15.625
105600 files
bucket 4 with min_duration=15.625 and max_duration=20.00001
136192 files
total: 521344

eval files: 2406
eval batch size: 8 (these are long form audio)

num gpus: 8 (for all the cases)
batch size: 1
bucketing batch size: 30
bucketing strategy: synced_randomized 
step per epoch (bucketing): 8966
training time per epoch: 1h15m

batch size: 30
step per epoch (non bucketing): 17417
training time per epoch: 1h45m

```
However, if I just train the original wav data set (same eval set)
```
total files: 521543
train batch size: 30
step per epoch: 2212
training time per epoch: 20m
```

**Reply by titu1994** [reply]
Look at your number of steps per epoch in the original dataset - 2212 and the number of files - 521543. Given your batch size of 30, yould need 521543 // 30 ~ 17,000 steps for 1 GPU and 1 node Assume you have 8 A100s on one node, that's 5215430/30/8 = 2173 steps. So this matches well. 

Now look at your bucketing dataset - 8966 steps per epoch - why is this number 4x the previous number of steps ?

Look at non bucketing (I assume tarred) - that's roughly 17k steps - that's the number for 1 GPU not 8. So something is possibly wrong in the GPU configuration - the number of GPUs selected is wrong

**Reply by titu1994** [reply]
Instead of using epochs with bucketing, try using fixed number of update steps. If you have 2200 update steps per epoch in the original dataset, use exactly 2200 * num epochs trainer.max_steps and see if the speed remains same or close to same.

**Reply by riqiang-dp** [reply]
I think bucketing should be compared to the non bucketing but tarred dataset, because it's also tarred. So it's actually less steps but non-bucketing is already 8x of the original.
But this is a good find! Sorry I didn't notice the relative difference between the step numbers. I will definitely look into it. Thanks! However when I did check the GPU usage, training with the tarred dataset seems normal, very high usage of GPU and memory for all the GPUs, the list of processes running also checks out.

**Reply by riqiang-dp** [reply]
Update: finally got time to look into this, turns out @titu1994 was absolutely right, the GPU was not configured correctly. I was doing some finetuning and the pre-trained model's `world_size` was not updated by using `set_trainer()`. I see that this is only the case with the version I was using, 1.7, where `set_trainer()` only updates `app_state.world_size` but not model's `world_size`. But `get_tarred_dataset()` relies on the `model.world_size` to set up distributed data loading.

I set the world size manually and now the training is much much faster, batch number is set correctly.

In the main branch this has been fixed, so I guess I should also update to the latest version of NeMo :)

Thanks @titu1994 and @VahidooX.

**Reply by titu1994** [reply]
Hmm so set trainer does not update model.world_size in <= NeMo 1.7? But model.world_size should automatically be from trainer.world_size..  interesting that such silent bugs cause huge performance loss. 

Well im glad you got correct speedup ! Hopefully bucketing helps training on large datasets

**Reply by riqiang-dp** [reply]
for your reference if it's of any use: 
<1.8.0:
https://github.com/NVIDIA/NeMo/blob/c16b89426d413da3b6ad848482674291df642720/nemo/core/classes/modelPT.py#L1241
main:
https://github.com/NVIDIA/NeMo/blob/a0e94df5e43e45ff28923e252cd9a1f3a4d9bd75/nemo/core/classes/modelPT.py#L1240

Yeah now it's so much faster for large datasets

**Reply by mehadi92** [reply]
@titu1994 is this issue solved in the latest version of Nemo? what will happen if we set  devices=-1 in the trainer? 

I'm getting slow training with data bucketing

**Reply by titu1994** [reply]
Devices -1 just uses all devices. It should be fast.

**Reply by mehadi92** [reply]
@titu1994 How num_steps are calculated for adaptive batch size for the bucketing dataset? should we use max_steps instead of max_epoch for tarred dataset + bucketing dataset?

**Reply by mehadi92** [reply]
@titu1994 

I have,  A100 40GB

For `without data bucketing`
Total audio file = 11740160
batch size = 24
num GPUs = 16
num_steps = 33203
to complete 1 epoch it takes = ~6 hr

For `with data bucketing`
Total audio file = 11740160
`batch size = 1`
`bucketing batch size = 26`
num GPUs = 16
`num_steps = 95190`
it shows me ~18 hr to complete 1 epoch. This time is taken when 2% of 1 epoch complete

Could not get any clue why `num_steps is 3x more` in data bucketing training

**Reply by titu1994** [reply]
For bucketing only max steps should be used.

**Reply by titu1994** [reply]
Please read the documentation here of how to setup bucketing - https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/asr/datasets.html#bucketing-datasets

**Reply by mehadi92** [reply]
> For bucketing only max steps should be used.

It would be great if this information is added to the Nemo documentation

**Reply by VahidooX** [reply]
Bucketing should work with both max_epochs and max_steps. Something is wrong here. The number of steps with bucketing should be smaller, not larger. Would you please try a list for bucketing_batch_size? something like [60,48,36,24] for 4 buckets.

**Reply by mehadi92** [reply]
@VahidooX, Thanks for your reply.

To compare bucketing and no-bucketing I try also try following,

with bucketing
data = 250 hr
max audio duration = 18.5
buckets = 3
shads = 64
bucketing batch size = [128, 64, 32]
** Above setup takes ~14 min to complete 1 epoch **

on the other hand `with no bucketing setup
batch size = 64
To complete 1 epoch it takes ~12 min

I use max_epochs in both cases
